authentication request specs

saneshark opened the pull request:
- this is not a full integration test, but it should easily be adapted to become one, or that's the goal.

saneshark commented:
@patrickvinograd @aub tagging you guys so you can see some of the progress on this. What I'm doing here is essentially creating a mechanism for recording integration specs similar to how one would do with capybara and then playing them back. You should be able to recreate the VCR cassettes locally doing some of the following:

add this to settings.local.yml

```
integration_recorder:
  enabled: true
  replay: false
  base_cassette_dir: spec/support/vcr_cassettes
  inbound_cassette_dir: integration/saml_loa3_user/inbound
  outbound_cassette_dir: integration/saml_loa3_user/outbound
```

This enables the initializer and rack middleware for recording.

- Inbound is characterized as vets-api / vets-website interactions, it happens as a Rack middleware.
- Outbound is characterized as backend related requests, hence why it is invoked in an around_action in application_controller.

In the rspec example you'll see I'm using an Episode struct, this just helps to facilitate using the recorded inbound interactions within the spec itself. I've had to use aggregate failures, because for whatever reason it appears as though fakeredis is not maintaining its in-memory store of User and Session between individual examples and I've had to run all the requests in a single example. Aggregate failures allows you to see each failed expectation within a failed example.

Let me know if you have any thoughts, recommendations, suggestions. Naturally specs are still failing, and most of this proabbly has to do with the timestamp freezing. I'm using the timestamp from the recorded cassette, but I think I need to use the timestamp that is used when saving the SAML assertion, alternately I could do some stubbing or VCR.casette filtering to make things a little more easier to manage. Eventually some of what you see in the rspec request spec will be moved into custom matchers and an improved DSL for handling or supporting this, but for now I just want to build something easy to record and manage.

saneshark commented:
I probably need to make this into a gem so that I can:

- test everything in isolation
- have the episodes stuff be an actual class
- add more helpful RSpec matchers
- provide better flexibility around whether or not you want randomization to be enabled or disabled
- have a unique README on how to use it

One alternative that recently occurred to me though would be to see if we can get a containerized version of third party identity provider, then it would be a matter of setting up a CI that runs selenium specs on the 3 containers.

aub commented:
This LGTM if you'll add a bunch of comments in the initializer where most of the complexity resides.

saneshark commented:
@aub I added comments. @omgitsbillryan would you mind reviewing when you have chance?

You can test it locally if you want. The settings in the initializer are straight-forward enough, just need to invoke it when running your rails server. I've updated the instructions above to no longer require Settings.yml

omgitsbillryan commented on spec/request/authentication/standard_authentication.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+  end
+
+  it 'does the tests', :aggregate_failures, :skip_mvi, :skip_emis do
```
`does the tests` is not very useful.  Might I recommend, 
```
Rspec.describe 'Authentication' ...
```
followed with,
```
it 'authenticates at loa3' do ...
```

saneshark replied:
Agree with these comments. I started working on a gem: 

https://github.com/saneshark/dual_deck

But I'm not going to work on that on company time for now. Right now I just needed a basic baseline to have some confidence mucking around with stuff. I'll continue to work on the gem on the weekend on my own time and eventually replace.
